### PR TITLE
Implement signatureHelp and implementation

### PR DESCRIPTION
**Pull request by icsaszar**

### CODE REVIEW

File: examples/server.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 use futures::future;
 use jsonrpc_core::{BoxFuture, Result};
+use lsp_types::request::GotoImplementationResponse;
```
Comment (ebkalderon) [inline examples/server.rs]: I think that this `use` statement should be combined with the existing `tower_lsp::lsp_types::request` import on line 5, similar to how it's done [here](https://github.com/ebkalderon/tower-lsp/blob/e2db3532d5d7b8dc38585e77d787f3743a49b845/src/lib.rs#L11).

End-users shouldn't have to depend on `lsp-types` directly in order to interact with this library, and can import from the re-exported symbols through `tower-lsp` in order to keep versions consistent.

File: examples/server.rs [hunk 2]
```diff
@@ -18,6 +19,8 @@ impl LanguageServer for Backend {
     type DefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
     type TypeDefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
     type HighlightFuture = BoxFuture<Option<Vec<DocumentHighlight>>>;
+    type SignatureHelpFuture = BoxFuture<Option<SignatureHelp>>;
+    type GotoImplementationFuture = BoxFuture<Option<GotoImplementationResponse>>;
```
Comment (ebkalderon) [inline examples/server.rs]: I would generally like to keep the ordering of associated types and methods consistent in the order they are described in the [current supported LSP specification](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/).

In this case:

* `GotoImplementationFuture` should be moved to immediately below `TypeDefinitionFuture`.
* `SignatureHelpFuture` should be moved to immediately below `HoverFuture`.

File: examples/server.rs [hunk 3]
```diff
@@ -125,6 +128,14 @@ impl LanguageServer for Backend {
     fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
         Box::new(future::ok(None))
     }
+
+    fn signature_help(&self, _: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
+        Box::new(future::ok(None))
+    }
+
+    fn goto_implementation(&self, _: TextDocumentPositionParams) -> Self::GotoImplementationFuture {
```
Comment (ebkalderon) [inline examples/server.rs]: This method should be moved to immediately below `LanguageServer::goto_type_definition()`.

File: examples/server.rs [hunk 4]
```diff
@@ -125,6 +128,14 @@ impl LanguageServer for Backend {
     fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
         Box::new(future::ok(None))
     }
+
+    fn signature_help(&self, _: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
```
Comment (ebkalderon) [inline examples/server.rs]: This method should be moved to immediately below `LanguageServer::hover()`.

File: src/delegate.rs [hunk 5]
```diff
@@ -98,6 +98,12 @@ pub trait LanguageServerCore {
 
     #[rpc(name = "textDocument/documentHighlight", raw_params)]
     fn document_highlight(&self, params: Params) -> BoxFuture<Option<Vec<DocumentHighlight>>>;
+
+    #[rpc(name = "textDocument/signatureHelp", raw_params)]
+    fn signature_help(&self, params: Params) -> BoxFuture<Option<SignatureHelp>>;
```
Comment (ebkalderon) [inline src/delegate.rs]: This method should be moved to immediately below `LanguageServerCore::hover()`.

File: src/delegate.rs [hunk 6]
```diff
@@ -98,6 +98,12 @@ pub trait LanguageServerCore {
...
+
+    #[rpc(name = "textDocument/implementation", raw_params)]
+    fn goto_implementation(&self, params: Params) -> BoxFuture<Option<GotoImplementationResponse>>;
```
Comment (ebkalderon) [inline src/delegate.rs]: This method should be moved to immediately below `LanguageServerCore::goto_type_definition()`.

File: src/delegate.rs [hunk 7]
```diff
@@ -270,6 +276,18 @@ impl<T: LanguageServer> LanguageServerCore for Delegate<T> {
             Box::new(self.server.document_highlight(p))
         })
     }
+
+    fn signature_help(&self, params: Params) -> BoxFuture<Option<SignatureHelp>> {
```
Comment (ebkalderon) [inline src/delegate.rs]: This method should be moved to immediately below `Delegate::hover()`.

File: src/delegate.rs [hunk 8]
```diff
@@ -270,6 +276,18 @@ impl<T: LanguageServer> LanguageServerCore for Delegate<T> {
...
+        self.delegate_request::<SignatureHelpRequest, _>(params, |p| {
+            Box::new(self.server.signature_help(p))
+        })
+    }
+
+    fn goto_implementation(&self, params: Params) -> BoxFuture<Option<GotoImplementationResponse>> {
```
Comment (ebkalderon) [inline src/delegate.rs]: This method should be moved to immediately below `Delegate::goto_type_definition()`.

File: src/lib.rs [hunk 9]
```diff
@@ -25,6 +25,9 @@
 //!     type DefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
 //!     type TypeDefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
 //!     type HighlightFuture = BoxFuture<Option<Vec<DocumentHighlight>>>;
+//!     type SignatureHelpFuture = BoxFuture<Option<SignatureHelp>>;
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be moved to immediately below `HoverFuture`.

File: src/lib.rs [hunk 10]
```diff
@@ -25,6 +25,9 @@
...
+//!     type GotoImplementationFuture = BoxFuture<Option<GotoImplementationResponse>>;
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be renamed to `ImplementationFuture` to be consistent with `TypeDefinitionFuture` and should be moved to immediately below `TypeDefinitionFuture`.

File: src/lib.rs [hunk 11]
```diff
@@ -344,6 +360,31 @@ pub trait LanguageServer: Send + Sync + 'static {
     ///
     /// [`textDocument/documentHighlight`]: https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#textDocument_documentHighlight
     fn document_highlight(&self, params: TextDocumentPositionParams) -> Self::HighlightFuture;
+
+    /// The [`textDocument/signatureHelp`] request is sent from the client to the server to request
+    /// signature information at a given cursor position.
+    ///
+    /// [`textDocument/signatureHelp`]: https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#textDocument_signatureHelp
+    fn signature_help(&self, params: TextDocumentPositionParams) -> Self::SignatureHelpFuture;
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `LanguageServer::hover()`.

File: src/lib.rs [hunk 12]
```diff
@@ -344,6 +360,31 @@ pub trait LanguageServer: Send + Sync + 'static {
...
+    /// field to `true` in the [`initialize`] method:
+    ///
+    /// ```text
+    /// InitializeParams::capabilities::text_document::implementation::link_support
+    /// ```
+    ///
+    /// [`textDocument/implementation`]: https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#textDocument_implementation
+    /// [`GotoImplementationResponse::Link`]: https://docs.rs/lsp-types/0.63.1/lsp_types/request/enum.GotoDefinitionResponse.html
+    /// [`initialize`]: #tymethod.initialize
+    fn goto_implementation(
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `LanguageServer::goto_type_definition()`.

File: src/lib.rs [hunk 13]
```diff
@@ -431,4 +474,15 @@ impl<S: ?Sized + LanguageServer> LanguageServer for Box<S> {
     fn document_highlight(&self, params: TextDocumentPositionParams) -> Self::HighlightFuture {
         (**self).document_highlight(params)
     }
+
+    fn signature_help(&self, params: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
+        (**self).signature_help(params)
+    }
+
+    fn goto_implementation(
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `LanguageServer::goto_type_definition()`.

File: src/service.rs [hunk 14]
```diff
@@ -235,6 +237,17 @@ mod tests {
         fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
             Box::new(future::ok(None))
         }
+
+        fn signature_help(&self, _: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
+            Box::new(future::ok(None))
+        }
+
+        fn goto_implementation(
```
Comment (ebkalderon) [inline src/service.rs]: This method should be moved to immediately below `LspService::goto_type_definition()`.

File: examples/server.rs (hunk 2 above)
Comment (ebkalderon) [reply examples/server.rs]: Additionally, to be consistent with the existing `TypeDefinitionFuture`, we should also rename this to be `ImplementationFuture`.

File: src/lib.rs [hunk 15]
```diff
@@ -431,4 +474,15 @@ impl<S: ?Sized + LanguageServer> LanguageServer for Box<S> {
     fn document_highlight(&self, params: TextDocumentPositionParams) -> Self::HighlightFuture {
         (**self).document_highlight(params)
     }
+
+    fn signature_help(&self, params: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `LanguageServer::hover()`.

File: src/service.rs [hunk 16]
```diff
@@ -235,6 +237,17 @@ mod tests {
         fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
             Box::new(future::ok(None))
         }
+
+        fn signature_help(&self, _: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
```
Comment (ebkalderon) [inline src/service.rs]: This method should be moved to immediately below `LspService::hover()`.

File: src/service.rs [hunk 17]
```diff
@@ -192,6 +192,8 @@ mod tests {
         type DefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
         type TypeDefinitionFuture = BoxFuture<Option<GotoDefinitionResponse>>;
         type HighlightFuture = BoxFuture<Option<Vec<DocumentHighlight>>>;
+        type SignatureHelpFuture = BoxFuture<Option<SignatureHelp>>;
```
Comment (ebkalderon) [inline src/service.rs]: This future should be moved to immediately below `HoverFuture`.

File: src/service.rs [hunk 18]
```diff
@@ -192,6 +192,8 @@ mod tests {
...
+        type GotoImplementationFuture = BoxFuture<Option<GotoImplementationResponse>>;
```
Comment (ebkalderon) [inline src/service.rs]: This future should be moved to immediately below `TypeDefinitionFuture`.

File: src/lib.rs [hunk 19]
```diff
@@ -356,6 +397,8 @@ impl<S: ?Sized + LanguageServer> LanguageServer for Box<S> {
     type DefinitionFuture = S::DefinitionFuture;
     type TypeDefinitionFuture = S::TypeDefinitionFuture;
     type HighlightFuture = S::HighlightFuture;
+    type SignatureHelpFuture = S::SignatureHelpFuture;
+    type GotoImplementationFuture = S::GotoImplementationFuture;
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be moved to immediately below `TypeDefinitionFuture`.

File: src/lib.rs [hunk 20]
```diff
@@ -356,6 +397,8 @@ impl<S: ?Sized + LanguageServer> LanguageServer for Box<S> {
     type DefinitionFuture = S::DefinitionFuture;
     type TypeDefinitionFuture = S::TypeDefinitionFuture;
     type HighlightFuture = S::HighlightFuture;
+    type SignatureHelpFuture = S::SignatureHelpFuture;
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be moved to immediately below `HoverFuture`.

File: src/lib.rs [hunk 21]
```diff
@@ -133,6 +144,11 @@ pub trait LanguageServer: Send + Sync + 'static {
     type TypeDefinitionFuture: Future<Item = Option<GotoDefinitionResponse>, Error = Error> + Send;
     /// Response returned when a document highlight action is requested.
     type HighlightFuture: Future<Item = Option<Vec<DocumentHighlight>>, Error = Error> + Send;
+    /// Response returned when a signature help action is requested.
+    type SignatureHelpFuture: Future<Item = Option<SignatureHelp>, Error = Error> + Send;
+    /// Response returned when a goto implementation action is requested.
+    type GotoImplementationFuture: Future<Item = Option<GotoImplementationResponse>, Error = Error>
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be renamed to `ImplementationFuture` to be consistent with `TypeDefinitionFuture` and moved to immediately below `TypeDefinitionFuture`.

File: src/lib.rs [hunk 22]
```diff
@@ -133,6 +144,11 @@ pub trait LanguageServer: Send + Sync + 'static {
     type TypeDefinitionFuture: Future<Item = Option<GotoDefinitionResponse>, Error = Error> + Send;
     /// Response returned when a document highlight action is requested.
     type HighlightFuture: Future<Item = Option<Vec<DocumentHighlight>>, Error = Error> + Send;
+    /// Response returned when a signature help action is requested.
+    type SignatureHelpFuture: Future<Item = Option<SignatureHelp>, Error = Error> + Send;
```
Comment (ebkalderon) [inline src/lib.rs]: This future should be moved to immediately below `SignatureHelpFuture`.

File: src/lib.rs [hunk 23]
```diff
@@ -69,6 +72,14 @@
 //!     fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
 //!         Box::new(future::ok(None))
 //!     }
+//!
+//!     fn signature_help(&self,params: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
+//!         Box::new(future::ok(None))
+//!     }
+//!
+//!     fn goto_implementation(&self,params: TextDocumentPositionParams) -> Self::GotoImplementationFuture {
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `Backend::goto_type_definition()`.

File: src/lib.rs [hunk 24]
```diff
@@ -69,6 +72,14 @@
 //!     fn document_highlight(&self, _: TextDocumentPositionParams) -> Self::HighlightFuture {
 //!         Box::new(future::ok(None))
 //!     }
+//!
+//!     fn signature_help(&self,params: TextDocumentPositionParams) -> Self::SignatureHelpFuture {
```
Comment (ebkalderon) [inline src/lib.rs]: This method should be moved to immediately below `Backend::hover()`.

**Comment by icsaszar** [comment]
Thanks for the feedback, I really appreciate that you care about small details like the order (which I totally agree with since there are a lot of methods and it would be quite hard to navigate without an IDE).

I think i can fix everything today.

**Comment by icsaszar** [comment]
@ebkalderon Everything should be fixed

**Comment by ebkalderon** [comment]
Affects #10.

**Pull request merged**